// jsdoctypeparser/peg_lib/jsdoctype.js
module.exports = (function() {
  /*
   * Generated by PEG.js 0.8.0.
   *
   * http://pegjs.majda.cz/
   */

  function peg$subclass(child, parent) {
    function ctor() { this.constructor = child; }
    ctor.prototype = parent.prototype;
    child.prototype = new ctor();
  }

  function SyntaxError(message, expected, found, offset, line, column) {
    this.message  = message;
    this.expected = expected;
    this.found    = found;
    this.offset   = offset;
    this.line     = line;
    this.column   = column;

    this.name     = "SyntaxError";
  }

  peg$subclass(SyntaxError, Error);

  function parse(input) {
    var options = arguments.length > 1 ? arguments[1] : {},

        peg$FAILED = {},

        peg$startRuleFunctions = { typeExpr: peg$parsetypeExpr },
        peg$startRuleFunction  = peg$parsetypeExpr,

        peg$c0 = peg$FAILED,
        peg$c1 = [],
        peg$c2 = function(prefixModifiersWithWhiteSpaces, modifieeWithWhiteSpaces, postfixModifiersWithWhiteSpaces) {
            var prefixModifiers = lodash.pluck(prefixModifiersWithWhiteSpaces, 0);
            var modifiee = modifieeWithWhiteSpaces;
            var postfixModifiers = lodash.pluck(postfixModifiersWithWhiteSpaces, 0);

            var modifiersOrderedByPriority = postfixModifiers.concat(reverse(prefixModifiers));
            var rootNode = modifiersOrderedByPriority.reduce(function(prevNode, operator) {
              switch (operator.operatorType) {
                case OperatorType.UNION:
                  var unionOperator = operator;
                  return {
                    type: NodeType.UNION,
                    left: prevNode,
                    right: unionOperator.right,
                  };
                case OperatorType.MEMBER:
                  var memberOperator = operator;
                  return {
                    type: NodeType.MEMBER,
                    owner: prevNode,
                    name: memberOperator.memberName,
                  };
                case OperatorType.INNER_MEMBER:
                  var memberOperator = operator;
                  return {
                    type: NodeType.INNER_MEMBER,
                    owner: prevNode,
                    name: memberOperator.memberName,
                  };
                case OperatorType.INSTANCE_MEMBER:
                  var memberOperator = operator;
                  return {
                    type: NodeType.INSTANCE_MEMBER,
                    owner: prevNode,
                    name: memberOperator.memberName,
                  };
                case OperatorType.GENERIC:
                  var genericOperator = operator;
                  return {
                    type: NodeType.GENERIC,
                    subject: prevNode,
                    objects: genericOperator.objects,
                  }
                case OperatorType.ARRAY:
                  var arrayOperator = operator;
                  return {
                    type: NodeType.GENERIC,
                    subject: {
                      type: NodeType.NAME,
                      name: 'Array'
                    },
                    objects: [ prevNode ],
                  };
                case OperatorType.OPTIONAL:
                  return {
                    type: NodeType.OPTIONAL,
                    value: prevNode,
                  };
                case OperatorType.NULLABLE:
                  return {
                    type: NodeType.NULLABLE,
                    value: prevNode,
                  };
                case OperatorType.NOT_NULLABLE:
                  return {
                    type: NodeType.NOT_NULLABLE,
                    value: prevNode,
                  };
                case OperatorType.VARIADIC:
                  return {
                    type: NodeType.VARIADIC,
                    value: prevNode,
                  };
                default:
                  throw Error('Unexpected token: ' + token);
              }
            }, modifiee);


            return rootNode;
          },
        peg$c3 = "(",
        peg$c4 = { type: "literal", value: "(", description: "\"(\"" },
        peg$c5 = ")",
        peg$c6 = { type: "literal", value: ")", description: "\")\"" },
        peg$c7 = function(wrapped) {
            return wrapped;
          },
        peg$c8 = function(name) {
            return {
              type: NodeType.NAME,
              name: name
            };
          },
        peg$c9 = /^[a-zA-Z_$]/,
        peg$c10 = { type: "class", value: "[a-zA-Z_$]", description: "[a-zA-Z_$]" },
        peg$c11 = /^[a-zA-Z0-9_$]/,
        peg$c12 = { type: "class", value: "[a-zA-Z0-9_$]", description: "[a-zA-Z0-9_$]" },
        peg$c13 = "module",
        peg$c14 = { type: "literal", value: "module", description: "\"module\"" },
        peg$c15 = ":",
        peg$c16 = { type: "literal", value: ":", description: "\":\"" },
        peg$c17 = function(filePath) {
            return {
              type: NodeType.MODULE,
              path: filePath
            };
          },
        peg$c18 = /^[a-zA-Z_0-9_$.\/\-]/,
        peg$c19 = { type: "class", value: "[a-zA-Z_0-9_$.\\/\\-]", description: "[a-zA-Z_0-9_$.\\/\\-]" },
        peg$c20 = "*",
        peg$c21 = { type: "literal", value: "*", description: "\"*\"" },
        peg$c22 = function() {
            return { type: NodeType.ANY };
          },
        peg$c23 = "?",
        peg$c24 = { type: "literal", value: "?", description: "\"?\"" },
        peg$c25 = function() {
            return { type: NodeType.UNKNOWN };
          },
        peg$c26 = "function",
        peg$c27 = { type: "literal", value: "function", description: "\"function\"" },
        peg$c28 = null,
        peg$c29 = function(paramParts, returnedTypePart) {
            var modifierGroups = lodash.groupBy(paramParts, lodash.property('modifierType'));

            var params = [];
            var noModifiers = modifierGroups[FunctionModifierType.NONE];
            if (noModifiers) {
              params = noModifiers.map(function(paramPartWithNoModifier) {
                return paramPartWithNoModifier.value;
              });
            }


            var thisValue = null;
            var thisValueModifiers = modifierGroups[FunctionModifierType.THIS];
            if (thisValueModifiers) {
              if (thisValueModifiers.length > 1) {
                reportSemanticIssue('"this" keyword should be declared only once');
              }

              // Enable the only first thisValue modifier.
              thisValue = thisValueModifiers[0].value;
            }


            var newValue = null;
            var newModifiers = modifierGroups[FunctionModifierType.NEW];
            if (newModifiers) {
              if (newModifiers.length > 1) {
                reportSemanticIssue('"new" keyword should be declared only once');
              }

              // Enable the only first new instance modifier.
              newValue = newModifiers[0].value;
            }

            var returnedTypeNode = returnedTypePart ? returnedTypePart[3] : null;

            return {
              type: NodeType.FUNCTION,
              params: params,
              returns: returnedTypeNode,
              thisValue: thisValue,
              newValue: newValue,
            };
          },
        peg$c30 = ",",
        peg$c31 = { type: "literal", value: ",", description: "\",\"" },
        peg$c32 = function(firstParam, restParamsWithComma) {
            var params = buildByFirstAndRest(firstParam, restParamsWithComma, 3);
            return params;
          },
        peg$c33 = "this",
        peg$c34 = { type: "literal", value: "this", description: "\"this\"" },
        peg$c35 = function(value) {
            return {
              modifierType: FunctionModifierType.THIS,
              value: value
            };
          },
        peg$c36 = "new",
        peg$c37 = { type: "literal", value: "new", description: "\"new\"" },
        peg$c38 = function(value) {
            return {
              modifierType: FunctionModifierType.NEW,
              value: value
            };
          },
        peg$c39 = function(value) {
            return {
              modifierType: FunctionModifierType.NONE,
              value: value
            };
          },
        peg$c40 = "{",
        peg$c41 = { type: "literal", value: "{", description: "\"{\"" },
        peg$c42 = "}",
        peg$c43 = { type: "literal", value: "}", description: "\"}\"" },
        peg$c44 = function(firstEntry, restEntriesWithComma) {
            var entries = buildByFirstAndRest(firstEntry, restEntriesWithComma, 3);

            return {
              type: NodeType.RECORD,
              entries: entries
            };
          },
        peg$c45 = function(key, valueWithColon) {
            var hasValue = Boolean(valueWithColon);
            var value = hasValue ? valueWithColon[3] : null;

            return {
              type: NodeType.RECORD_ENTRY,
              key: key,
              value: value,
              hasValue: hasValue,
            };
          },
        peg$c46 = function() {
            return { operatorType: OperatorType.NULLABLE };
          },
        peg$c47 = "!",
        peg$c48 = { type: "literal", value: "!", description: "\"!\"" },
        peg$c49 = function() {
            return { operatorType: OperatorType.NOT_NULLABLE };
          },
        peg$c50 = "=",
        peg$c51 = { type: "literal", value: "=", description: "\"=\"" },
        peg$c52 = function() {
            return { operatorType: OperatorType.OPTIONAL };
          },
        peg$c53 = "...",
        peg$c54 = { type: "literal", value: "...", description: "\"...\"" },
        peg$c55 = function() {
            return { operatorType: OperatorType.VARIADIC };
          },
        peg$c56 = function(right) {
            return {
              operatorType: OperatorType.UNION,
              right: right,
            };
          },
        peg$c57 = "|",
        peg$c58 = { type: "literal", value: "|", description: "\"|\"" },
        peg$c59 = function(memberName) {
            return {
              operatorType: OperatorType.MEMBER,
              memberName: memberName,
            };
          },
        peg$c60 = ".",
        peg$c61 = { type: "literal", value: ".", description: "\".\"" },
        peg$c62 = function(name) {
            return name;
          },
        peg$c63 = function(memberName) {
            return {
              operatorType: OperatorType.INNER_MEMBER,
              memberName: memberName,
            };
          },
        peg$c64 = "~",
        peg$c65 = { type: "literal", value: "~", description: "\"~\"" },
        peg$c66 = function(memberName) {
            return {
              operatorType: OperatorType.INSTANCE_MEMBER,
              memberName: memberName,
            };
          },
        peg$c67 = "#",
        peg$c68 = { type: "literal", value: "#", description: "\"#\"" },
        peg$c69 = function(objects) {
            return {
              operatorType: OperatorType.GENERIC,
              objects: objects,
            };
          },
        peg$c70 = ".<",
        peg$c71 = { type: "literal", value: ".<", description: "\".<\"" },
        peg$c72 = "<",
        peg$c73 = { type: "literal", value: "<", description: "\"<\"" },
        peg$c74 = ">",
        peg$c75 = { type: "literal", value: ">", description: "\">\"" },
        peg$c76 = function(first, restsWithComma) {
            var objects = buildByFirstAndRest(first, restsWithComma, 3);
            return objects;
          },
        peg$c77 = "[]",
        peg$c78 = { type: "literal", value: "[]", description: "\"[]\"" },
        peg$c79 = function() {
            return { operatorType: OperatorType.ARRAY };
          },
        peg$c80 = /^[ \t\r\n ]/,
        peg$c81 = { type: "class", value: "[ \\t\\r\\n ]", description: "[ \\t\\r\\n ]" },

        peg$currPos          = 0,
        peg$reportedPos      = 0,
        peg$cachedPos        = 0,
        peg$cachedPosDetails = { line: 1, column: 1, seenCR: false },
        peg$maxFailPos       = 0,
        peg$maxFailExpected  = [],
        peg$silentFails      = 0,

        peg$result;

    if ("startRule" in options) {
      if (!(options.startRule in peg$startRuleFunctions)) {
        throw new Error("Can't start parsing from rule \"" + options.startRule + "\".");
      }

      peg$startRuleFunction = peg$startRuleFunctions[options.startRule];
    }

    function text() {
      return input.substring(peg$reportedPos, peg$currPos);
    }

    function offset() {
      return peg$reportedPos;
    }

    function line() {
      return peg$computePosDetails(peg$reportedPos).line;
    }

    function column() {
      return peg$computePosDetails(peg$reportedPos).column;
    }

    function expected(description) {
      throw peg$buildException(
        null,
        [{ type: "other", description: description }],
        peg$reportedPos
      );
    }

    function error(message) {
      throw peg$buildException(message, null, peg$reportedPos);
    }

    function peg$computePosDetails(pos) {
      function advance(details, startPos, endPos) {
        var p, ch;

        for (p = startPos; p < endPos; p++) {
          ch = input.charAt(p);
          if (ch === "\n") {
            if (!details.seenCR) { details.line++; }
            details.column = 1;
            details.seenCR = false;
          } else if (ch === "\r" || ch === "\u2028" || ch === "\u2029") {
            details.line++;
            details.column = 1;
            details.seenCR = true;
          } else {
            details.column++;
            details.seenCR = false;
          }
        }
      }

      if (peg$cachedPos !== pos) {
        if (peg$cachedPos > pos) {
          peg$cachedPos = 0;
          peg$cachedPosDetails = { line: 1, column: 1, seenCR: false };
        }
        advance(peg$cachedPosDetails, peg$cachedPos, pos);
        peg$cachedPos = pos;
      }

      return peg$cachedPosDetails;
    }

    function peg$fail(expected) {
      if (peg$currPos < peg$maxFailPos) { return; }

      if (peg$currPos > peg$maxFailPos) {
        peg$maxFailPos = peg$currPos;
        peg$maxFailExpected = [];
      }

      peg$maxFailExpected.push(expected);
    }

    function peg$buildException(message, expected, pos) {
      function cleanupExpected(expected) {
        var i = 1;

        expected.sort(function(a, b) {
          if (a.description < b.description) {
            return -1;
          } else if (a.description > b.description) {
            return 1;
          } else {
            return 0;
          }
        });

        while (i < expected.length) {
          if (expected[i - 1] === expected[i]) {
            expected.splice(i, 1);
          } else {
            i++;
          }
        }
      }

      function buildMessage(expected, found) {
        function stringEscape(s) {
          function hex(ch) { return ch.charCodeAt(0).toString(16).toUpperCase(); }

          return s
            .replace(/\\/g,   '\\\\')
            .replace(/"/g,    '\\"')
            .replace(/\x08/g, '\\b')
            .replace(/\t/g,   '\\t')
            .replace(/\n/g,   '\\n')
            .replace(/\f/g,   '\\f')
            .replace(/\r/g,   '\\r')
            .replace(/[\x00-\x07\x0B\x0E\x0F]/g, function(ch) { return '\\x0' + hex(ch); })
            .replace(/[\x10-\x1F\x80-\xFF]/g,    function(ch) { return '\\x'  + hex(ch); })
            .replace(/[\u0180-\u0FFF]/g,         function(ch) { return '\\u0' + hex(ch); })
            .replace(/[\u1080-\uFFFF]/g,         function(ch) { return '\\u'  + hex(ch); });
        }

        var expectedDescs = new Array(expected.length),
            expectedDesc, foundDesc, i;

        for (i = 0; i < expected.length; i+
